File: CSharp/One/Ast/Statements.cs
using One.Ast;
using System.Collections.Generic;
using System.Linq;

namespace One.Ast
{
    public class Statement : IHasAttributesAndTrivia, IAstNode {
        public string leadingTrivia { get; set; }
        public Dictionary<string, string> attributes { get; set; }
        
        public Statement()
        {
            
        }
    }
    
    public class IfStatement : Statement {
        public Expression condition;
        public Block then;
        public Block else_;
        
        public IfStatement(Expression condition, Block then, Block else_): base()
        {
            this.condition = condition;
            this.then = then;
            this.else_ = else_;
        }
    }
    
    public class ReturnStatement : Statement {
        public Expression expression;
        
        public ReturnStatement(Expression expression): base()
        {
            this.expression = expression;
        }
    }
    
    public class ThrowStatement : Statement {
        public Expression expression;
        
        public ThrowStatement(Expression expression): base()
        {
            this.expression = expression;
        }
    }
    
    public class ExpressionStatement : Statement {
        public Expression expression;
        
        public ExpressionStatement(Expression expression): base()
        {
            this.expression = expression;
        }
    }
    
    public class BreakStatement : Statement {
        
    }
    
    public class ContinueStatement : Statement {
        
    }
    
    public class UnsetStatement : Statement {
        public Expression expression;
        
        public UnsetStatement(Expression expression): base()
        {
            this.expression = expression;
        }
    }
    
    public class VariableDeclaration : Statement, IVariableWithInitializer, IReferencable {
        public string name { get; set; }
        public IType type { get; set; }
        public Expression initializer { get; set; }
        public List<VariableDeclarationReference> references;
        public MutabilityInfo mutability { get; set; }
        
        public VariableDeclaration(string name, IType type, Expression initializer): base()
        {
            this.name = name;
            this.type = type;
            this.initializer = initializer;
            this.references = new List<VariableDeclarationReference>();
        }
        
        public Reference createReference()
        {
            return new VariableDeclarationReference(this);
        }
    }
    
    public class WhileStatement : Statement {
        public Expression condition;
        public Block body;
        
        public WhileStatement(Expression condition, Block body): base()
        {
            this.condition = condition;
            this.body = body;
        }
    }
    
    public class DoStatement : Statement {
        public Expression condition;
        public Block body;
        
        public DoStatement(Expression condition, Block body): base()
        {
            this.condition = condition;
            this.body = body;
        }
    }
    
    public class ForeachVariable : IVariable, IReferencable {
        public string name { get; set; }
        public IType type { get; set; }
        public List<ForeachVariableReference> references;
        public MutabilityInfo mutability { get; set; }
        
        public ForeachVariable(string name)
        {
            this.name = name;
            this.references = new List<ForeachVariableReference>();
        }
        
        public Reference createReference()
        {
            return new ForeachVariableReference(this);
        }
    }
    
    public class ForeachStatement : Statement {
        public ForeachVariable itemVar;
        public Expression items;
        public Block body;
        
        public ForeachStatement(ForeachVariable itemVar, Expression items, Block body): base()
        {
            this.itemVar = itemVar;
            this.items = items;
            this.body = body;
        }
    }
    
    public class ForVariable : IVariableWithInitializer, IReferencable {
        public string name { get; set; }
        public IType type { get; set; }
        public Expression initializer { get; set; }
        public List<ForVariableReference> references;
        public MutabilityInfo mutability { get; set; }
        
        public ForVariable(string name, IType type, Expression initializer)
        {
            this.name = name;
            this.type = type;
            this.initializer = initializer;
            this.references = new List<ForVariableReference>();
        }
        
        public Reference createReference()
        {
            return new ForVariableReference(this);
        }
    }
    
    public class ForStatement : Statement {
        public ForVariable itemVar;
        public Expression condition;
        public Expression incrementor;
        public Block body;
        
        public ForStatement(ForVariable itemVar, Expression condition, Expression incrementor, Block body): base()
        {
            this.itemVar = itemVar;
            this.condition = condition;
            this.incrementor = incrementor;
            this.body = body;
        }
    }
    
    public class CatchVariable : IVariable, IReferencable {
        public string name { get; set; }
        public IType type { get; set; }
        public List<CatchVariableReference> references;
        public MutabilityInfo mutability { get; set; }
        
        public CatchVariable(string name, IType type)
        {
            this.name = name;
            this.type = type;
            this.references = new List<CatchVariableReference>();
        }
        
        public Reference createReference()
        {
            return new CatchVariableReference(this);
        }
    }
    
    public class TryStatement : Statement {
        public Block tryBody;
        public CatchVariable catchVar;
        public Block catchBody;
        public Block finallyBody;
        
        public TryStatement(Block tryBody, CatchVariable catchVar, Block catchBody, Block finallyBody): base()
        {
            this.tryBody = tryBody;
            this.catchVar = catchVar;
            this.catchBody = catchBody;
            this.finallyBody = finallyBody;
            if (this.catchBody == null && this.finallyBody == null)
                throw new Error("try without catch and finally is not allowed");
        }
    }
    
    public class Block {
        public List<Statement> statements;
        
        public Block(Statement[] statements)
        {
            this.statements = statements.ToList();
        }
    }
}

File: PHP/One/Ast/Statements.php
<?php

namespace One\Ast\Statements;

use One\Ast\Types\IVariableWithInitializer;
use One\Ast\Types\IVariable;
use One\Ast\Types\IHasAttributesAndTrivia;
use One\Ast\Types\IAstNode;
use One\Ast\Types\MutabilityInfo;
use One\Ast\Expressions\Expression;
use One\Ast\References\ForVariableReference;
use One\Ast\References\ForeachVariableReference;
use One\Ast\References\VariableDeclarationReference;
use One\Ast\References\IReferencable;
use One\Ast\References\Reference;
use One\Ast\References\CatchVariableReference;
use One\Ast\Interfaces\IType;

class Statement implements IHasAttributesAndTrivia, IAstNode {
    public $leadingTrivia;
    public $attributes;
    
    function __construct() {
        
    }
}

class IfStatement extends Statement {
    public $condition;
    public $then;
    public $else_;
    
    function __construct($condition, $then, $else_) {
        parent::__construct();
        $this->condition = $condition;
        $this->then = $then;
        $this->else_ = $else_;
    }
}

class ReturnStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class ThrowStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class ExpressionStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class BreakStatement extends Statement {
    
}

class ContinueStatement extends Statement {
    
}

class UnsetStatement extends Statement {
    public $expression;
    
    function __construct($expression) {
        parent::__construct();
        $this->expression = $expression;
    }
}

class VariableDeclaration extends Statement implements IVariableWithInitializer, IReferencable {
    public $name;
    public $type;
    public $initializer;
    public $references;
    public $mutability;
    
    function __construct($name, $type, $initializer) {
        parent::__construct();
        $this->name = $name;
        $this->type = $type;
        $this->initializer = $initializer;
        $this->references = array();
    }
    
    function createReference() {
        return new VariableDeclarationReference($this);
    }
}

class WhileStatement extends Statement {
    public $condition;
    public $body;
    
    function __construct($condition, $body) {
        parent::__construct();
        $this->condition = $condition;
        $this->body = $body;
    }
}

class DoStatement extends Statement {
    public $condition;
    public $body;
    
    function __construct($condition, $body) {
        parent::__construct();
        $this->condition = $condition;
        $this->body = $body;
    }
}

class ForeachVariable implements IVariable, IReferencable {
    public $name;
    public $type;
    public $references;
    public $mutability;
    
    function __construct($name) {
        $this->name = $name;
        $this->references = array();
    }
    
    function createReference() {
        return new ForeachVariableReference($this);
    }
}

class ForeachStatement extends Statement {
    public $itemVar;
    public $items;
    public $body;
    
    function __construct($itemVar, $items, $body) {
        parent::__construct();
        $this->itemVar = $itemVar;
        $this->items = $items;
        $this->body = $body;
    }
}

class ForVariable implements IVariableWithInitializer, IReferencable {
    public $name;
    public $type;
    public $initializer;
    public $references;
    public $mutability;
    
    function __construct($name, $type, $initializer) {
        $this->name = $name;
        $this->type = $type;
        $this->initializer = $initializer;
        $this->references = array();
    }
    
    function createReference() {
        return new ForVariableReference($this);
    }
}

class ForStatement extends Statement {
    public $itemVar;
    public $condition;
    public $incrementor;
    public $body;
    
    function __construct($itemVar, $condition, $incrementor, $body) {
        parent::__construct();
        $this->itemVar = $itemVar;
        $this->condition = $condition;
        $this->incrementor = $incrementor;
        $this->body = $body;
    }
}

class CatchVariable implements IVariable, IReferencable {
    public $name;
    public $type;
    public $references;
    public $mutability;
    
    function __construct($name, $type) {
        $this->name = $name;
        $this->type = $type;
        $this->references = array();
    }
    
    function createReference() {
        return new CatchVariableReference($this);
    }
}

class TryStatement extends Statement {
    public $tryBody;
    public $catchVar;
    public $catchBody;
    public $finallyBody;
    
    function __construct($tryBody, $catchVar, $catchBody, $finallyBody) {
        parent::__construct();
        $this->tryBody = $tryBody;
        $this->catchVar = $catchVar;
        $this->catchBody = $catchBody;
        $this->finallyBody = $finallyBody;
        if ($this->catchBody === null && $this->finallyBody === null)
            throw new \OneLang\Error("try without catch and finally is not allowed");
    }
}

class Block {
    public $statements;
    
    function __construct($statements) {
        $this->statements = $statements;
    }
}
